firebase-components/src/main/java/com/google/firebase/components/ComponentRuntime.java

```diff
@@ -35,35 +38,86 @@
   private final Map<Class<?>, Provider<?>> lazyInstanceMap = new HashMap<>();
-  private final Map<Class<?>, Provider<Set<?>>> lazySetMap = new HashMap<>();
+  private final Map<Class<?>, LazySet<?>> lazySetMap = new HashMap<>();
+  private final List<Provider<ComponentRegistrar>> unprocessedRegistrarProviders;
   private final EventBus eventBus;
+  private Boolean eagerComponentsInitializedWith = null;
 
   /**
    * Creates an instance of {@link ComponentRuntime} for the provided {@link ComponentRegistrar}s
    * and any additional components.
+   *
+   * @deprecated Use {@link #create(Executor, Iterable, Component[])} instead.
    */
+  @Deprecated
   public ComponentRuntime(
       Executor defaultEventExecutor,
       Iterable<ComponentRegistrar> registrars,
       Component<?>... additionalComponents) {
+    this(defaultEventExecutor, toProviders(registrars), true, additionalComponents);
+  }
+
+  /**
+   * Creates an instance of {@link ComponentRuntime} for the provided {@link ComponentRegistrar}s
+   * and any additional components.
+   */
+  public static ComponentRuntime create(
+      Executor defaultEventExecutor,
+      Iterable<Provider<ComponentRegistrar>> registrars,
+      Component<?>... additionalComponents) {
+    return new ComponentRuntime(defaultEventExecutor, registrars, true, additionalComponents);
+  }
+
+  private ComponentRuntime(
+      Executor defaultEventExecutor,
+      Iterable<Provider<ComponentRegistrar>> registrars,
+      boolean disambiguateConstructorOverload,
+      Component<?>... additionalComponents) {
     eventBus = new EventBus(defaultEventExecutor);
+
     List<Component<?>> componentsToAdd = new ArrayList<>();
-    componentsToAdd.add(Component.of(eventBus, EventBus.class, Subscriber.class, Publisher.class));
 
-    for (ComponentRegistrar registrar : registrars) {
-      componentsToAdd.addAll(registrar.getComponents());
-    }
+    componentsToAdd.add(Component.of(eventBus, EventBus.class, Subscriber.class, Publisher.class));
+    componentsToAdd.add(Component.of(this, ComponentLoader.class));
     for (Component<?> additionalComponent : additionalComponents) {
       if (additionalComponent != null) {
         componentsToAdd.add(additionalComponent);
       }
     }
 
-    CycleDetector.detect(componentsToAdd);
+    unprocessedRegistrarProviders = iterableToList(registrars);
+
+    discoverComponents(componentsToAdd);
+  }
+
+  private synchronized void discoverComponents(List<Component<?>> componentsToAdd) {
+
+    Iterator<Provider<ComponentRegistrar>> iterator = unprocessedRegistrarProviders.iterator();
+    while (iterator.hasNext()) {
+      Provider<ComponentRegistrar> provider = iterator.next();
+      try {
+        ComponentRegistrar registrar = provider.get();
+        if (registrar != null) {
+          componentsToAdd.addAll(registrar.getComponents());
+          iterator.remove();
+        }
+      } catch (InvalidRegistrarException ex) {
```

Comment:
The most common case is when we do the initial discovery, the class `FooRegistrar` is not found, so we keep the provider around for later use. Once module foo is loaded, we can suddenly load `FooRegistrar` but it's not a valid registrar(see discoverLazy's javadoc for why it can be invalid). wdyt?